### PR TITLE
Fix issues in spring deployment

### PR DESCRIPTION
**Pull request by Flanker32**
- Waiting for instances up and show url after deployment, for #4197 
- Use copy constructor when generate run state, for #4205 
- Show error dialog when spring cloud app deploy failed
- Change Runtime Version in deploy model to string for serialize

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeployConfiguration.java [hunk 1]
```diff
@@ -166,8 +172,8 @@ public void setCreateNewApp(boolean isNewApp) {
         springCloudModel.setCreateNewApp(isNewApp);
     }
 
-    public void setRuntimeVersion(RuntimeVersion runtimeVersion) {
-        springCloudModel.setRuntimeVersion(runtimeVersion);
+    public void saveRuntimeVersion(RuntimeVersion runtimeVersion) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeployConfiguration.java]: null check

Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeployConfiguration.java]: Objects.toString

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeployConfiguration.java [hunk 2]
```diff
@@ -107,7 +113,7 @@ public String getAppName() {
     }
 
     public RuntimeVersion getRuntimeVersion() {
-        return springCloudModel.getRuntimeVersion();
+        return RuntimeVersion.fromString(springCloudModel.getRuntimeVersion());
     }
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeployConfiguration.java]: null check for springCloudModel.getRuntimeVersion ?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java [hunk 3]
```diff
@@ -100,4 +131,69 @@ protected String getDeployTarget() {
...
+            processHandler.setText("Failed to get the deployment status, you may get the status in portal later.");
+        }
+    }
+
+    @FunctionalInterface
+    private interface SupplierWithIOException<T> {
+        T get() throws IOException;
+    }
+
+    private <T> T getResourceWithTimeout(SupplierWithIOException<T> consumer, Predicate<T> predicate,
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java]: static

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java [hunk 4]
```diff
@@ -100,4 +131,69 @@ protected String getDeployTarget() {
...
+                                           this::isDeploymentDone, GET_STATUS_TIMEOUT, TimeUnit.SECONDS);
+            processHandler.setText(
+                    "Deployment done with status " + deploymentResourceInner.properties().status().toString());
+        } catch (Exception e) {
+            processHandler.setText("Failed to get the deployment status, you may get the status in portal later.");
+        }
+    }
+
+    @FunctionalInterface
+    private interface SupplierWithIOException<T> {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java]: static?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeployConfiguration.java (hunk 2 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeployConfiguration.java]: Added empty string check

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeployConfiguration.java (hunk 1 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeployConfiguration.java]: Fixed

**Pull request merged**